File: 2016/day2/pb.lua
local current = { x = 3, y = 1 }
local pad = {
	{ nil, nil, "1", nil, nil },
	{ nil, "2", "3", "4", nil },
	{ "5", "6", "7", "8", "9" },
	{ nil, "A", "B", "C", nil },
	{ nil, nil, "D", nil, nil },
}

local function is_valid(x, y)
	return pad[x] and pad[x][y] ~= nil
end

for line in io.lines("./input.txt") do
	for i = 1, #line do
		local char = line:sub(i, i)

		local new_x, new_y = current.x, current.y

		if char == "U" then
			new_x = current.x - 1
		elseif char == "D" then
			new_x = current.x + 1
		elseif char == "L" then
			new_y = current.y - 1
		elseif char == "R" then
			new_y = current.y + 1
		end

		if is_valid(new_x, new_y) then
			current.x = new_x
			current.y = new_y
		end
	end

	io.write(pad[current.x][current.y])
end

print()

File: 2016-py/day2/pb.py
def is_valid(x, y, pad):
    return 0 <= x < len(pad) and 0 <= y < len(pad[0]) and pad[x][y] is not None


def main():
    current = [2, 0]  # starts on 5

    pad = [
        [None, None, "1", None, None],
        [None, "2", "3", "4", None],
        ["5", "6", "7", "8", "9"],
        [None, "A", "B", "C", None],
        [None, None, "D", None, None],
    ]

    with open("input.txt", "r") as fin:
        for line in fin:
            line = line.strip()

            for char in line:
                new_x, new_y = current

                if char == "U":
                    new_x -= 1
                elif char == "D":
                    new_x += 1
                elif char == "L":
                    new_y -= 1
                elif char == "R":
                    new_y += 1

                if is_valid(new_x, new_y, pad):
                    current = [new_x, new_y]

            print(pad[current[0]][current[1]], end="")

    print()


if __name__ == "__main__":
    main()

